File: android/app/src/main/java/com/example/smartwords/data/WordStore.kt
// Word model + store. Shared by the app and the Glance widget.
// Bundled seed asset is the source of words (831 entries). Selection is a pure
// function of the clock + rotation speed, mirroring ios/Shared/WordStore.swift.

package com.example.smartwords.data

import android.content.Context
import org.json.JSONObject
import java.io.File
import java.util.Calendar
import java.util.concurrent.TimeUnit

// Lexica design fields (ipa/synonyms/antonyms/origin) are optional — the seed has
// none yet, so detail screens render those sections only when present.
data class Word(
    val word: String,
    val definition: String,
    val short: String? = null,   // 2-3 word gloss for the widget; falls back to definition
    val pos: String? = null,     // part of speech
    val example: String? = null, // usage sentence
    val ipa: String? = null,
    val synonyms: List<String>? = null,
    val antonyms: List<String>? = null,
    val origin: String? = null,
)

object WordStore {
    const val DEFAULT_ROTATION_HOURS = 3

    // Cached so the word list is parsed once per process.
    @Volatile private var cached: List<Word>? = null

    /** Drop the in-process cache so the next read re-parses (after a batch fetch). */
    fun invalidate() { cached = null }

    /** Today's words: the fetched daily batch if cached, else the bundled seed. Never empty. */
    fun words(ctx: Context): List<Word> {
        cached?.let { return it }
        val parsed = cachedBatch(ctx)
            ?: runCatching { parse(readAsset(ctx)) }.getOrNull()?.takeIf { it.isNotEmpty() }
            ?: listOf(Word("smartwords", "a word a day, on your home screen.", short = "word a day", pos = "noun"))
        cached = parsed
        return parsed
    }

    /** The daily batch written by [BatchWorker], or null if absent/invalid. */
    private fun cachedBatch(ctx: Context): List<Word>? {
        val f = File(ctx.filesDir, "batch.json")
        if (!f.exists()) return null
        return runCatching { parse(f.readText()) }.getOrNull()?.takeIf { it.isNotEmpty() }
    }

    /** Parse a batch/seed JSON string into words, or empty on any failure. Used by the worker to validate. */
    fun parseOrEmpty(json: String): List<Word> = runCatching { parse(json) }.getOrDefault(emptyList())

    fun batchFile(ctx: Context): File = File(ctx.filesDir, "batch.json")

    /**
     * Index of the word for a given moment — pure function of clock + rotation.
     * Advances one slot every [rotationHours] and a full day's worth per day, so it
     * walks the whole list and never repeats the same day's set.
     */
    fun index(words: List<Word>, rotationHours: Int, cal: Calendar = Calendar.getInstance()): Int {
        if (words.isEmpty()) return 0
        val hours = rotationHours.coerceIn(1, 24)
        val slotsPerDay = 24 / hours
        // dayOfEra: whole days since epoch — equivalent to iOS's ordinality(of:.day,in:.era).
        val dayOfEra = (cal.timeInMillis / TimeUnit.DAYS.toMillis(1)).toInt()
        val slot = cal.get(Calendar.HOUR_OF_DAY) / hours
        val raw = (dayOfEra.toLong() * slotsPerDay + slot) % words.size
        return ((raw + words.size) % words.size).toInt()
    }

    fun word(words: List<Word>, rotationHours: Int, cal: Calendar = Calendar.getInstance()): Word =
        words[index(words, rotationHours, cal)]

    private fun readAsset(ctx: Context): String =
        ctx.assets.open("seed-words.json").bufferedReader().use { it.readText() }

    private fun parse(json: String): List<Word> {
        val arr = JSONObject(json).getJSONArray("words")
        return (0 until arr.length()).map { i ->
            val o = arr.getJSONObject(i)
            Word(
                word = o.getString("word"),
                definition = o.getString("definition"),
                short = o.optString("short").ifEmpty { null },
                pos = o.optString("pos").ifEmpty { null },
                example = o.optString("example").ifEmpty { null },
                ipa = o.optString("ipa").ifEmpty { null },
                synonyms = o.optJSONArray("synonyms")?.let { a -> (0 until a.length()).map { a.getString(it) } },
                antonyms = o.optJSONArray("antonyms")?.let { a -> (0 until a.length()).map { a.getString(it) } },
                origin = o.optString("origin").ifEmpty { null },
            )
        }
    }
}

File: ios/Shared/WordStore.swift
// Shared between the app and the widget (included in both targets via project.yml).
// The widget renders the current word and rotates by the clock; the app reads the
// same store to show expanded detail and to change the rotation speed.

import Foundation

// MARK: - Model

struct Word: Codable, Hashable {
    let word: String
    let definition: String
    let short: String?     // 2-3 word gloss for the widget; falls back to definition
    let pos: String?       // optional: not all sources provide part of speech
    let example: String?   // optional usage sentence
    // Lexica design fields — rendered only when present (seed has none yet).
    let ipa: String?
    let synonyms: [String]?
    let antonyms: [String]?
    let origin: String?
}

struct Batch: Codable { let date: String?; let words: [Word] }

// MARK: - Store (App Group cache + bundled seed fallback)

enum WordStore {
    // App Group shared by the app + widget. Works on the simulator with the
    // entitlement declared in project.yml.
    static let appGroup = "group.com.example.smartwords"
    static let defaultRotationHours = 3

    private static var shared: UserDefaults? { UserDefaults(suiteName: appGroup) }

    /// How often the widget advances, in hours. User-set (App Group), else default.
    /// Always a divisor of 24 so a whole number of slots fits in a day.
    static var rotationHours: Int {
        let v = shared?.integer(forKey: "rotationHours") ?? 0
        return v > 0 ? v : defaultRotationHours
    }
    static func setRotationHours(_ hours: Int) { shared?.set(hours, forKey: "rotationHours") }

    /// Cache a fetched daily batch (validated JSON) for the widget + app to read.
    static func cache(batchData: Data) { shared?.set(batchData, forKey: "batch") }
    static var cachedBatchDate: String? {
        guard let data = shared?.data(forKey: "batch") else { return nil }
        return (try? JSONDecoder().decode(Batch.self, from: data))?.date
    }

    /// Today's words: cached batch if present, else bundled seed. Never empty.
    static func words() -> [Word] {
        if let data = shared?.data(forKey: "batch"),
           let batch = try? JSONDecoder().decode(Batch.self, from: data),
           !batch.words.isEmpty {
            return batch.words
        }
        return seed()
    }

    private static func seed() -> [Word] {
        guard let url = Bundle.main.url(forResource: "seed-words", withExtension: "json"),
              let data = try? Data(contentsOf: url),
              let batch = try? JSONDecoder().decode(Batch.self, from: data)
        else { return [Word(word: "smartwords",
                            definition: "a word a day, on your home screen.",
                            short: "word a day",
                            pos: "noun", example: nil,
                            ipa: nil, synonyms: nil, antonyms: nil, origin: nil)] }
        return batch.words
    }

    /// Index of the word for a given moment — pure function of clock + batch.
    /// Advances one slot every `rotationHours` and one full day's worth per day,
    /// so it walks the whole list and never repeats the same day's set.
    /// rotation/calendar are injectable so the logic is unit-testable.
    /// ponytail: sequential walk; for non-repeating order, shuffle the batch server-side.
    static func index(at date: Date, in words: [Word],
                      rotationHours hours: Int = rotationHours,
                      calendar cal: Calendar = .current) -> Int {
        guard !words.isEmpty else { return 0 }
        let slotsPerDay = 24 / hours
        let day = cal.ordinality(of: .day, in: .era, for: date) ?? 0
        let slot = cal.component(.hour, from: date) / hours
        return (day * slotsPerDay + slot) % words.count
    }

    static func word(at date: Date, in words: [Word]) -> Word {
        words[index(at: date, in: words)]
    }
}
